Implement Filter/TryFilter

spadafiva opened the pull request.

spadafiva committed:
Adds filter and try filter implementations

spadafiva committed:
Implement Filter

spadafiva committed:
Remove @testable declaration

spadafiva committed:
Fix linting

spadafiva commented:
@broadwaylamb could you take a look at this one as well?

spadafiva committed:
Updates tests and creates testing helper

spadafiva commented on Tests/OpenCombineTests/Helpers/TestHelper.swift:
```diff
@@ -0,0 +1,52 @@
+#endif
+
+@available(macOS 10.15, *)
+class TestHelper<Value: Equatable,
```
There is a lot of really similar boilerplate in the tests for setting up the custom subscriptions, custom publisher and tracking subscribers.  This makes setting up a test case as simple as passing a closure to create the chained subscriber.  Let me know what you think @broadwaylamb .  If not exactly this, something along these lines would help remove a lot of the duplicate code in the testing suites.

broadwaylamb replied:
I think it's a great idea!

Can you add documentation for this helper so it's clear what this class is for and how it can be used?

Let's just indent generic parameters below to be aligned with `Value`, and rename `SUT` to `Sut` to match existing usages of this name.

```swift
class TestHelper<Value: Equatable,
                 Failure,
                 SourceValue: Equatable,
                 SourcePublisher,
                 Sut: Publisher>
    where Sut.Failure == Failure,
          Sut.Output == Value,
          SourcePublisher: CustomPublisherBase<SourceValue>
{
```

Also, `TestHelper` is too generic name, maybe something like OperatorTestHelper will do?

Besides, since we have these same-type constraints `Sut.Failure == Failure` and `Sut.Output == Value`, do we actually need `Value` and `Failure` generic parameters? We can just use type aliases:

```swift
typealias Value = Sut.Output
typealias Failure = Sut.Failure
```

spadafiva replied:
👍

spadafiva commented on Tests/OpenCombineTests/Helpers/TrackingSubscriber.swift:
```diff
@@ -87,6 +87,12 @@ final class TrackingSubscriberBase<Value: Equatable,
     private let _receiveCompletion: ((Subscribers.Completion<Failure>) -> Void)?
     private let _onDeinit: (() -> Void)?
 
+    var onSubscribe: ((Subscription) -> Void)?
```
These are some hooks to make it easier to observe changes on a tracking subscriber.  I'm mostly using the `onSubscribe` in the TestHelper init since self can't be captured until after init

broadwaylamb replied:
maybe it's better make existing properties like `_receiveSubscription` internal instead of adding a new property?

spadafiva replied:
I've intentionally left the value which returns a demand as a separate property.  In other projects I've worked on, it's helpful to have something which can inspect what's going on without having to return the correct values.  I think having the closure based observers makes it easier to write more targeted tests against specific behaviors without having to worry about changing the preconditions that are set when the underlying TrackingSubscriber is initialized.  What do you think?

spadafiva committed:
Fix allTests to include all tests

spadafiva committed:
Renames TestHelper to OperatorTestHelper and adds documentation

spadafiva committed:
Adds more test coverage

spadafiva commented:
@broadwaylamb There's 100% test coverage and I've addressed all the feedback.  I'd also like to be able to refactor to use the OperatorTestHelper in more upcoming PRs.  Let me know what/if anything else needs to be done.

broadwaylamb commented on Sources/OpenCombine/Publishers/Publishers.Filter.swift:
```diff
@@ -0,0 +1,180 @@
+        super.init(downstream: downstream)
+    }
+
+    let description: String
```
Storing `description` is probably not the way to go. Combine uses two different subclasses for `Filter` and `TryFilter`, which share logic in a common base class. It’s possible to make sure of it using `dump(type(of: subscription))` call.

Subclassing also allows us to avoid force casts below.

You can refer to `Map`/`TryMap` implementation to see how we do this.

spadafiva committed:
Updates to use subclasses for filter / tryfilter

broadwaylamb commented on Sources/OpenCombine/Publishers/Publishers.Filter.swift:
```diff
@@ -0,0 +1,208 @@
+
+    func receive(subscription: Subscription) {
+        upstreamSubscription = subscription
+        subscription.request(.unlimited)
```
Can you please add a test to check if Apple's Combine also requests `.unlimited` from the upstream upon subscription?

It seems that there is something less straightforward going on: https://forums.swift.org/t/a-uicontrol-event-publisher-example/26215/18

Almost all operators (except `sink` and maybe something else) in Combine are lazy, which means that they don't immediately consume their input, processing it on-demand instead.

spadafiva replied:
@broadwaylamb I’ve added a test and it passing for me locally in compatibility mode.

spadafiva replied:
@broadwaylamb I’ve added a test and it passing for me locally in compatibility mode.

spadafiva committed:
Adds subscription test

broadwaylamb commented on Tests/OpenCombineTests/PublisherTests/FilterTests.swift:
```diff
@@ -0,0 +1,187 @@
+        }
+
+        XCTAssertEqual(helper.subscription.history, [.requested(.unlimited)])
+    }
```
The reason it passes like this is because `OperatorTestHelper` always requests `.unlimited`.

If we change this test to

```swift
func testFilterSubscriptionDemand() {
    let helper = OperatorTestHelper(publisherType: CustomPublisher.self,
                                    trackingDemand: .max(3),
                                    receiveValueDemand: .none) {
                                        $0.filter { $0.isMultiple(of: 2) }
                                    }

    XCTAssertEqual(helper.publisher.send(1), .max(1))
    XCTAssertEqual(helper.publisher.send(2), .max(0))
    XCTAssertEqual(helper.publisher.send(3), .max(1))
    XCTAssertEqual(helper.publisher.send(4), .max(0))
    XCTAssertEqual(helper.publisher.send(5), .max(1))
    XCTAssertEqual(helper.publisher.send(6), .max(0))
    XCTAssertEqual(helper.publisher.send(7), .max(1))
    XCTAssertEqual(helper.publisher.send(8), .max(0))

    XCTAssertEqual(helper.subscription.history, [.requested(.max(3))])
}
```

it will pass in compatibility mode, but fail with OpenCombine. The same goes for the test below.

I would suggest to adjust all tests to not request `.unlimited`, since it's not very representative.

Maybe it would be even better for `OperatorTestHelper` initializer arguments to not take default values, instead taking them explicitly to avoid this kind of problems.

It's important to test our publishers with different demands.

broadwaylamb replied:
You can look at my attempt to handle this in `CompactMap` implementation (#32), which is practically the same as `Filter`.

spadafiva committed:
Fix subscriber demand to be lazy

spadafiva committed:
Merge branch 'master' into add-filter-try-filter

spadafiva committed:
Fix CustomPublisherBase changes from master

spadafiva committed:
Fix iOS availability on test helper

Commit:
Updates availability for test functions

spadafiva commented:
@broadwaylamb I've fixed the issues and tests and added tests as per CompactMap.  What else needs to happen for this one?

broadwaylamb commented on Sources/OpenCombine/Publishers/Publishers.Filter.swift:
```diff
@@ -0,0 +1,213 @@
+        case .success(let isIncluded):
+            if isIncluded {
+                let newDemand = downstream.receive(input)
+                _unsatisfiedDemand += newDemand - 1
```
If `newDemand` is `.max(0)`, this will crash as subtracting from zero demand is not permitted.

Better use two in-place mutations here:

```swift
_unsatisfiedDemand -= 1
_unsatisfiedDemand += newDemand
```

And add a test of course :)

broadwaylamb commented on Sources/OpenCombine/Publishers/Publishers.Filter.swift:
```diff
@@ -0,0 +1,213 @@
+    private var _unsatisfiedDemand: Subscribers.Demand = .none
+    private var _isFinished = false
+
+    var errorTransform: (Upstream.Failure) -> Downstream.Failure
```
Why is this property needed?

spadafiva replied:
This allows use to use this in `receive` so that we don't have to write out a different "receive" for every subclass that is using _Filter and instead maps the error.  Otherwise we have an issues with generic typing that you need to force cast the Failure as! Downstream.Failure.

broadwaylamb commented on Sources/OpenCombine/Publishers/Publishers.Filter.swift:
```diff
@@ -0,0 +1,213 @@
+//  Created by Joseph Spadafora on 7/3/19.
+//
+
+import Foundation
```
We don’t want to depend on Foundation, let’s remove this import

spadafiva replied:
Ooops, I think that came with the boilerplate.  Good catch

broadwaylamb reviewed (changes requested):
I’m sorry for the delay, been quite a busy week.

broadwaylamb committed:
Merge branch 'master' into add-filter-try-filter

broadwaylamb committed:
Simplify Filter implementation, add more tests

broadwaylamb committed:
Ensure test suite consistency on Darwin and Linux

broadwaylamb committed:
Add missing tests to XCTestManifests.swift
